lib/gps.js

```diff
@@ -0,0 +1,418 @@
+ *
+ */
+
+function GPS(opts) {
+
+  var breakout, receiver, chip, state;
+
+  if (!(this instanceof GPS)) {
+    return new GPS(opts);
+  }
+
+  // Allow users to pass in a 2 element array for tx and rx pins
+  if (Array.isArray(opts)) {
+    opts = { pins: {tx: opts[0], rx: opts[1], onOff: opts[2] }};
+  }
+
+  if (typeof opts.pins === "undefined") {
+    opts.pins = {};
+  }
+
+  Board.Component.call(
+    this, opts = Board.Options(opts)
+  );
+
+
+
+  // Get user values for breakout, receiver and chip
+  breakout = opts.breakout || {};
+  receiver = opts.receiver;
+  chip = opts.chip;
+
+  // If a breakout is defined check for receiver and chip
+  if (Breakouts[breakout]) {
+    if (!receiver && Breakouts[breakout].receiver) {
+      receiver = Breakouts[breakout].receiver.value;
+    }
+
+    if (!chip && Breakouts[breakout].chip) {
+      chip = Breakouts[breakout].chip.value;
+    }
+  }
+
+  // If a receiver was defined or derived but chip was not
+  if (receiver && !chip && Receivers[receiver].chip) {
+    chip = Receivers[receiver].chip.value;
+  }
+
+  // Default case
+  if (!chip) {
+    chip = "DEFAULT";
+  }
+
+  // Allow users to pass in custom chip types
+  chip = typeof chip === "string" ?
+    Chips[chip] : opts.chip;
+
+  // Allow users to pass in custom receiver types
+  receiver = typeof receiver === "string" ?
+    Receivers[receiver] : opts.receiver;
+
+  // Chip decorates the instance
+  Object.defineProperties(this, chip);
+
+  // Receiver decorates this instance
+  if (receiver) {
+    Object.defineProperties(this, receiver);
+  }
+
+  // breakout decorates the instance
+  if (opts.breakout) {
+    breakout = typeof opts.breakout === "string" ?
+      Breakouts[opts.breakout] : opts.breakout;
+
+    Board.Controller.call(this, breakout, opts);
+  }
+
+  // If necessary set default property values
+  this.precision = opts.precision || 9;
+  this.baud = opts.baud || breakout.baud || chip.baud;
+
+  // Create a "state" entry for privately
+  // storing the state of the instance
+  state = {
+    sat: {},
+    pos: {},
+    nav: {},
+    frequency: 1000,
+    lowPowerMode: false
+  };
+
+  priv.set(this, state);
+
+  // Getters for private state values
+  Object.defineProperties(this, {
+    pos: {
+      get: function() {
+        return state.pos;
+      }
+    },
+    sat: {
+      get: function() {
+        return state.sat;
+      }
+    },
+    nav: {
+      get: function() {
+        return state.nav;
+      }
+    },
+    time: {
+      get: function() {
+        return state.time;
+      }
+    }
+  });
+
+  if (this.initialize) {
+    this.initialize(opts);
+  }
+
+}
+
+util.inherits(GPS, events.EventEmitter);
+
+/*
+ * Default intialization for serial GPS
+ */
+GPS.prototype.initialize = function(opts) {
+
+  var state = priv.get(this);
+  state.serialPort = opts.serialPort || opts.portId || opts.port || opts.bus || this.io.SERIAL_PORT_IDs.SW_SERIAL0 || this.io.SERIAL_PORT_IDs[0];
```

Comment:
I think that will require a change in firmata because in firmata io.SERIAL_PORT_IDs[0] ===  io.SERIAL_PORT_IDs.HW_SERIAL0 which we would never want as the default